### ISSUE TITLE
Possible bug on validation during training

### ISSUE THREAD

**Issue by kwang-greenhand** [NifTK/NiftyNet#66]
It seems I may find a bug on validation during training settings. 

What I observed: 
when using validation_every_n, I obtained two loss curves, one for training, one for validation. These were very close to each other, which made me feel confident on m model. However results on inference sets were very terrible. So I went back to the code of `application_driver.py`. Everything seems fine: `iter_message` is initialized, and `vars_to_run` is updated for iteration based on whether it's training or validation. 

What I did: 
I modified the run_vars function as follows:
    def run_vars(self, sess, message):
        """
        Running a TF session by retrieving variables/operations to run,
        along with data for feed_dict.

        This function sets ``message._current_iter_output`` with session.run
        outputs.
        """
        # update iteration status before the batch process
        self.app.set_iteration_update(message)
        collected = self.outputs_collector
        # building a dictionary of variables
        vars_to_run = message.ops_to_run
        

#############################################################
#### modified
        tf.logging.info('current phase is %s ', message.phase)
        try:
            tf.logging.info('gradient op is %s',vars_to_run['gradients'])
        except:
            tf.logging.info('gradient op doesnt exist')
        print('vars_to_run is ')
        print(vars_to_run)
#############################################################
####         
        
        if message.is_training:
            # always apply the gradient op during training
            vars_to_run['gradients'] = self.app.gradient_op
            
#############################################################
#### modified            
        else:
            del vars_to_run['gradients']
   
        print('Now vars_to_run is ')
        print(vars_to_run)
#############################################################

I can see the gradient_op exists even for validation iteration, and only disappear after explicitly deleting it. After this modification, the two loss curves were not matched any more... Which is good in terms of proving my point, and bad that my model still doesn't work...
